python/ray/tune/trial.py

```diff
@@ -0,0 +1,159 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import sys
+import traceback
+
+import ray
+
+from ray.rllib.agents import get_agent_class
+
+
+class Trial(object):
+    """A trial object holds the state for one model training run.
+
+    Trials are themselves managed by the TrialRunner class, which implements
+    the event loop for submitting trial runs to a Ray cluster.
+
+    Trials start in the PENDING state, and transition to RUNNING once started.
+    On error it transitions to ERROR, otherwise TERMINATED on success.
+    """
+
+    PENDING = 'PENDING'
+    RUNNING = 'RUNNING'
+    TERMINATED = 'TERMINATED'
+    ERROR = 'ERROR'
+
+    def __init__(
+            self, env_creator, alg, config={}, local_dir='/tmp/ray',
+            agent_id=None, resources={'cpu': 1}, stopping_criterion={},
+            checkpoint_freq=sys.maxsize, restore_path=None):
+        """Initialize a new trial.
+
+        The args here take the same meaning as the command line flags defined
+        in ray.tune.config_parser.
+        """
+
+        # Immutable config
+        self.env_creator = env_creator
+        if type(env_creator) is str:
+            self.env_name = env_creator
+        else:
+            self.env_name = "custom"
+        self.alg = alg
+        self.config = config
+        self.local_dir = local_dir
+        self.agent_id = agent_id
+        self.resources = resources
+        self.stopping_criterion = stopping_criterion
+        self.checkpoint_freq = checkpoint_freq
+        self.restore_path = restore_path
+
+        # Local trial state that is updated during the run
+        self.last_result = None
+        self.checkpoint_path = None
+        self.agent = None
+        self.status = Trial.PENDING
+
+    def start(self):
+        """Starts this trial.
+
+        If an error is encountered when starting the trial, an exception will
+        be thrown.
+        """
+
+        self.status = Trial.RUNNING
+        agent_cls = get_agent_class(self.alg)
+        cls = ray.remote(num_gpus=self.resources.get('gpu', 0))(agent_cls)
+        self.agent = cls.remote(
+            self.env_creator, self.config, self.local_dir,
```

Comment:
(don't know where to leave this) but just pointing out this does not support S3 directories.